pilot/pkg/serviceregistry/kube/controller/endpoints.go

```diff
@@ -178,25 +182,42 @@ func (e *endpointsController) onEvent(curr interface{}, event model.Event) error
 	return processEndpointEvent(e.c, e, ep.Name, ep.Namespace, event, curr)
 }
 
+func (e *endpointsController) forgetEndpoint(endpoint interface{}) {
+	ep := endpoint.(*v1.Endpoints)
+	key := kube.KeyFunc(ep.Name, ep.Namespace)
+	for _, ss := range ep.Subsets {
+		for _, ea := range ss.Addresses {
+			e.c.pods.dropNeedsUpdate(key, ea.IP)
+		}
+	}
+}
+
 func (e *endpointsController) buildIstioEndpoints(endpoint interface{}, host host.Name) []*model.IstioEndpoint {
 	endpoints := make([]*model.IstioEndpoint, 0)
 	ep := endpoint.(*v1.Endpoints)
+	key := kube.KeyFunc(ep.Name, ep.Namespace)
 	for _, ss := range ep.Subsets {
 		for _, ea := range ss.Addresses {
 			pod := e.c.pods.getPodByIP(ea.IP)
 			if pod == nil {
 				// This means, the endpoint event has arrived before pod event. This might happen because
 				// PodCache is eventually consistent. We should try to get the pod from kube-api server.
 				if ea.TargetRef != nil && ea.TargetRef.Kind == "Pod" {
-					pod = e.c.pods.getPod(ea.TargetRef.Name, ea.TargetRef.Namespace)
-					if pod == nil {
+					// There is a small chance getInformer may have the pod, but it hasn't made its way to the PodCache yet
+					// This is due to the shared queue. Try the getInformer store first
+					podFromInformer, f, err := e.c.pods.informer.GetStore().GetByKey(key)
+					if err != nil || !f {
 						// If pod is still not available, this an unusual case.
 						endpointsWithNoPods.Increment()
```

Comment:
I think we should probably remove this stat and logs below

Reply:
As I did in my PR - may be we should have syncQueueSize gauge metric that tells how many endpoints currently are missing pods?